src/ck-core/ckrdma.C

```diff
@@ -2093,3 +2059,510 @@ void zcPupIssueRgets(CmiUInt8 id, CkLocMgr *locMgr) {
+    }
+
+  } else { // found, perform rget
+
+    CkNcpyBufferPost post = iter->second;
+
+    if(CkPerformRget(post, destBuffer, destSize)) {
+      CmiLock(CksvAccess(_nodeZCPostReqLock));
+      CksvAccess(ncpyPostedReqNodeMap).erase(iter);
+      CmiUnlock(CksvAccess(_nodeZCPostReqLock));
+    }
+  }
+}
+
+// PE-level Match method
+void CkMatchBuffer(CkNcpyBufferPost *post, int index, int tag) {
+
+  post[index].postAsync = true;
+
+  // check if tag exists in posted buffer table
+  auto iter = CkpvAccess(ncpyPostedBufferMap).find(tag);
+
+  if(iter == CkpvAccess(ncpyPostedBufferMap).end()) {
+
+    auto iter2 = CkpvAccess(ncpyPostedReqMap).find(tag);
+
+    if(iter2 == CkpvAccess(ncpyPostedReqMap).end()) { // not found, insert into ncpyPostedReqMap
+      post[index].tag = tag;
+      CkpvAccess(ncpyPostedReqMap).emplace(post[index].tag, post[index]);
+    } else {
+      CkAbort("CkMatchBuffer: tag %d already exists, use another tag!\n", tag);
+    }
+
+  } else { // found, perform rget
+
+    CkPostedBuffer *buff = &(iter->second);
+    post[index].tag = tag;
+
+    if(CkPerformRget((post[index]), buff->buffer, buff->bufferSize)) {
+      CkpvAccess(ncpyPostedBufferMap).erase(iter);
+    }
+  }
+}
+
+// Node-level Match method
+void CkMatchNodeBuffer(CkNcpyBufferPost *post, int index, int tag) {
+
+  post[index].postAsync = true;
+
+  // check if tag exists posted buffer node table
+  auto iter = CksvAccess(ncpyPostedBufferNodeMap).find(tag);
+
+  if(iter == CksvAccess(ncpyPostedBufferNodeMap).end()) {
+
+    auto iter2 = CksvAccess(ncpyPostedReqNodeMap).find(tag);
+
+    if(iter2 == CksvAccess(ncpyPostedReqNodeMap).end()) { // not found, insert into ncpyPostedReqNodeMap
+      post[index].tag = tag;
+      CmiLock(CksvAccess(_nodeZCPostReqLock));
+      CksvAccess(ncpyPostedReqNodeMap).emplace(post[index].tag, post[index]);
+      CmiUnlock(CksvAccess(_nodeZCPostReqLock));
+    } else {
+      CkAbort("CkMatchNodeBuffer: tag %d already exists, use another tag!\n", tag);
+    }
+
+  } else { // found, perform rget
+
+    CkPostedBuffer *buff = &(iter->second);
+    post[index].tag = tag;
+
+    if(CkPerformRget((post[index]), buff->buffer, buff->bufferSize)) {
+      CmiLock(CksvAccess(_nodeZCBufferReqLock));
+      CksvAccess(ncpyPostedBufferNodeMap).erase(iter);
+      CmiUnlock(CksvAccess(_nodeZCBufferReqLock));
+    }
+  }
+}
+
+void initPostStruct(CkNcpyBufferPost *ncpyPost, int index) {
+  ncpyPost[index].regMode = CK_BUFFER_REG;
+  ncpyPost[index].deregMode = CK_BUFFER_DEREG;
+  ncpyPost[index].index = index;
+  ncpyPost[index].postAsync = false;
+}
+
+void setPostStruct(CkNcpyBufferPost *ncpyPost, int index, CkNcpyBuffer &buffObj, CmiUInt8 elemIndex) {
+  ncpyPost[index].srcBuffer = (void *)buffObj.ptr;
+  ncpyPost[index].srcSize = buffObj.cnt;
+  ncpyPost[index].ncpyEmInfo->tagArray = buffObj.ncpyEmInfo->tagArray;
+  ncpyPost[index].opIndex = index;
+  ncpyPost[index].arrayIndex = elemIndex;
+}
```

Comment:
It might be good to factor out the array indexing into `CkNcpyBufferPost & post = ncpyPost[index];`

Reply:
Fixed.